### PR TITLE
Allow permtations in Hadamard/Sum/Partial Trace specifications

### PR DESCRIPTION
**Pull request by under-Peter**

### CODE REVIEW

File: test/EinRule.jl [hunk 1]
```diff
@@ -82,4 +83,4 @@ using OMEinsum: match_rule, PairWise, Sum, Tr, DefaultRule,
     @test match_rule(Identity, ((1,2,3),), (1,2,3))
     @test !match_rule(Identity, ((1,2,3),), (1,3,2))
```
Comment (GiggleLiu) [inline test/EinRule.jl]: Except `match_rule`, please also test `einsum` is correctly implemented?

File: src/einsum.jl [hunk 2]
```diff
@@ -45,8 +45,9 @@ function einsum(::PTrace, ::EinCode{ixs,iy}, xs, size_dict) where {ixs, iy}
     tensortrace(xs[1], ixs[1], iy)
 end
 
-function einsum(::Hadamard, ::EinCode, xs, size_dict)
-    broadcast(*, xs...)
+function einsum(::Hadamard, ::EinCode{ixs, iy}, xs, size_dict) where {ixs, iy}
```
Comment (GiggleLiu) [inline src/einsum.jl]: For operations as simple as Hadamard test, could you please show the benchmark? I am afraid this implementation with permutation costs allocations, and will not be faster.

File: src/einsum.jl [hunk 3]
```diff
@@ -60,7 +61,10 @@ end
 
 function einsum(sm::Sum, code::EinCode{ixs, iy}, xs, size_dict) where {ixs, iy}
     dims = (findall(i -> i ∉ iy, ixs[1])...,)
-    dropdims(sum(xs[1], dims=dims), dims=dims)
+    (ix1,) = ixs
+    ix1f = filter!(i -> i in iy, collect(ix1))
+    perm = map(i -> findfirst(==(i), ix1f), iy)
+    permutedims(dropdims(sum(xs[1], dims=dims), dims=dims), perm)
```
Comment (GiggleLiu) [inline src/einsum.jl]: Also benchmark `Sum` please.

File: src/einsum.jl (hunk 2 above)
Comment (under-Peter) [reply src/einsum.jl]: You're right, compared to #60, this isn't an advantage:
```
ein"ijk,jki,ikj -> kji"(a,a,a)

with a = randn(100,100,100)
PR62 13ms 30Mib
master 19ms 7Mib
PR60 10ms 7Mib

with a = randn(10,10,10)
PR62 10μs 33KiB
master 12μs 8KiB
PR60 9μs 9KiB
```
I'll remove the Hadamard-modification then (regular broadcasting without permutation is still faster so without permutation, a hadamard-product should still be dispatched to `.*`)

File: src/einsum.jl (hunk 3 above)
Comment (under-Peter) [reply src/einsum.jl]: permuted sum is much faster for large arrays and about the same (if not a bit worse) for small ones:
```
#=
ein"ijk, ->kj"(a)

with a = randn(100,100,100)
PR62 470μs 160KiB
master 2ms 80KiB
PR60 1.3ms 80KiB

with a = randn(10,10,10)
PR62 3.5μs 2.63KiB
master 3.2μs 1.22KiB
PR60 3.1μs 1.44KiB
=#
Given those results, I'd keep the permutation&sum implementation, do you agree?

Comment (GiggleLiu) [reply src/einsum.jl]: Yes, :+1:

**Pull request merged**